cmd/sync/azure.go

```diff
@@ -0,0 +1,176 @@
+package main
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/compute/mgmt/compute"
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/network/mgmt/network"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	yaml "gopkg.in/yaml.v2"
+)
+
+// AzureClient allows you to get the list of IP addresses of VirtualMachines of a VirtualMachine Scale Set. It implements the CloudProvider interface
+type AzureClient struct {
+	config      *azureConfig
+	vmssClient  compute.VirtualMachineScaleSetsClient
+	ifaceClient network.InterfacesClient
+	authorizer  autorest.Authorizer
+}
+
+// NewAzureClient creates an AzureClient
+func NewAzureClient(data []byte) (*AzureClient, error) {
+	azureClient := &AzureClient{}
+	cfg, err := azureClient.parseAzureConfig(data)
+	if err != nil {
+		return nil, fmt.Errorf("error validating config: %v", err)
+	}
+
+	azureClient.config = cfg
+
+	err = azureClient.configure()
+	if err != nil {
+		return nil, fmt.Errorf("error configuring Azure Client: %v", err)
+	}
+
+	return azureClient, nil
+}
+
+// parseAzureConfig parses and validates AzureClient config
+func (client *AzureClient) parseAzureConfig(data []byte) (*azureConfig, error) {
+	cfg := &azureConfig{}
+	err := yaml.Unmarshal(data, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	err = validateAzureConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+// GetPrivateIPsForScalingGroup returns the list of IP addresses of instances of the Virtual Machine Scale Set
+func (client *AzureClient) GetPrivateIPsForScalingGroup(name string) ([]string, error) {
+	var ips []string
+
+	ctx := context.Background()
+
+	for iFaces, err := client.ifaceClient.ListVirtualMachineScaleSetNetworkInterfaces(ctx, client.config.ResourceGroupName, name); iFaces.NotDone() || err != nil; err = iFaces.NextWithContext(ctx) {
+		if err != nil {
+			return nil, err
+		}
+
+		for _, iFace := range iFaces.Values() {
+			if iFace.VirtualMachine != nil && iFace.VirtualMachine.ID != nil {
+				for _, n := range *iFace.IPConfigurations {
+					if *n.Primary && n.InterfaceIPConfigurationPropertiesFormat.PrivateIPAddress != nil {
```

Comment:
I think once the VirtualMachine.ID is != nil, Primary and IPConfigurations can't be nil (eg they are filled). But I can add the checks just in case, sure.

Reply:
Also, I think `IPConfigurations` is an slice of pointers `[]*SomeStruct` , so the for loop will prevent the panic because it won't iterate over it, right?